src/main/java/org/jenkinsci/plugins/artifactdeployer/MatrixArtifactDeployerPublisher.java

```diff
@@ -0,0 +1,60 @@
+package org.jenkinsci.plugins.artifactdeployer;
+
+import hudson.Extension;
+import hudson.Launcher;
+import hudson.matrix.MatrixAggregatable;
+import hudson.matrix.MatrixAggregator;
+import hudson.matrix.MatrixBuild;
+import hudson.matrix.MatrixConfiguration;
+import hudson.matrix.MatrixProject;
+import hudson.matrix.MatrixRun;
+import hudson.model.AbstractBuild;
+import hudson.model.AbstractProject;
+import hudson.model.BuildListener;
+import org.kohsuke.stapler.DataBoundConstructor;
+
+import java.io.IOException;
+import java.util.List;
+
+/**
+ * Extends ArtifactDeployerPublisher to support Matrix projects and implement MatrixAggregatable.
+ * @author <a href="mailto:nicolas.deloof@gmail.com">Nicolas De Loof</a>
+ */
+public class MatrixArtifactDeployerPublisher extends ArtifactDeployerPublisher implements MatrixAggregatable {
+
+    @DataBoundConstructor
+    public MatrixArtifactDeployerPublisher(List<ArtifactDeployerEntry> deployedArtifact, boolean deployEvenBuildFail) {
+        super(deployedArtifact, deployEvenBuildFail);
+    }
+
+    @Override
+    public boolean perform(AbstractBuild<?, ?> build, Launcher launcher, BuildListener listener) throws InterruptedException, IOException {
+        if (build.getProject() instanceof MatrixConfiguration) return true;
+        return super.perform(build, launcher, listener);
+    }
+
+    public MatrixAggregator createAggregator(final MatrixBuild build, final Launcher launcher, final BuildListener listener) {
+        return new MatrixAggregator(build, launcher, listener) {
+
+            @Override
+            public boolean endRun(MatrixRun run) throws InterruptedException, IOException {
+                boolean result = MatrixArtifactDeployerPublisher.super.perform((AbstractBuild) run, launcher, listener);
+                run.save();
+                return result;
+            }
+
+        };
+    }
+
+    @Extension(optional = true)
```

Comment:
Then it probably works by a detail of the JVM implementation. The enclosing type of a static class is not generally loaded when the nested class is loaded, since in the bytecode they are scarcely related. Some aspect of the verification of `DescriptorImpl` might be causing `MatrixArtifactDeployerPublisher` to be loaded transitively; or perhaps the error is caught during the call to its constructor when `[Descriptor.]clazz` is defined.

IOW, I know that loading the extension will ultimately fail; the question is when the error is thrown: when `optional=true` the Guice extension loader will reduce a stack trace to a one-line warning, but only if it is the one to catch the error.